### PR TITLE
Refactoring/gcp/iam

### PR DESCRIPTION
**Pull request by Remi05**
Migrated GCP's IAM service to the new architecture.  

Note that `gcp/resources/projects.py` and `gcp/facade/base.py` are the same as in the [Stackdriver logging PR](https://github.com/nccgroup/ScoutSuite/pull/229); they are simply shared by all services and will appear in all GCP refactoring PRs until one of the services is merged.

### CODE REVIEW

**Comment by Remi05** [comment]
The service seems to work fine but the resource counts have changed; In the previous implementation, IAM showed a count of 8 resources where as the new implementation shows 4 resources. In both cases, there are 4 service accounts, which is shown correctly in the top navigation bar for both implementations. I thought it might be due to the keys and bindings counts but they do not seem to be set, even in the previous implementation.

Before:
iam_beforerefactoring

After:
iam_afterrefactoring

**Comment by Remi05** [comment]
The build doesn't pass for Python 3.5 because of formatted strings using the `f'Text {my_variable}'` syntax which was introduced in Python 3.6. Do we still plan on supporting Python 3.5 or only 3.6 and 3.7?

**Comment by j4v** [comment]
> Do we still plan on supporting Python 3.5 or only 3.6 and 3.7?

Well asyncio is available from 3.5+, but 3.6+ should be the standard 3.x on most distros. If there's a benefit to only supporting 3.6+ that's something we could consider. I'm seeing a lot of changes happening in python 3.x so if we're along for the ride (and everyone will have to be from 2020 onward) then we might as well just accept that having outdated versions of python is not acceptable anymore.

**Comment by Aboisier** [comment]
> > Do we still plan on supporting Python 3.5 or only 3.6 and 3.7?
> 
> Well asyncio is available from 3.5+, but 3.6+ should be the standard 3.x on most distros. If there's a benefit to only supporting 3.6+ that's something we could consider. I'm seeing a lot of changes happening in python 3.x so if we're along for the ride (and everyone will have to be from 2020 onward) then we might as well just accept that having outdated versions of python is not acceptable anymore.

I'm not sure formatted string literals is a good enough reason to deprecate 3.5... but at the same time, it seems like there's barely anyone using 3.5. 

image

File: ScoutSuite/providers/gcp/configs/services.py [hunk 1]
```diff
@@ -20,11 +22,14 @@ def __init__(self, metadata=None, thread_config=4, projects=None, **kwargs):
...
 
+        gcp_facade = GCPFacade()
+        iam_facade = IAMFacade()
+
         self.cloudresourcemanager = CloudResourceManager(thread_config=thread_config)
         self.cloudstorage = CloudStorageConfig(thread_config=thread_config)
         self.cloudsql = CloudSQLConfig(thread_config=thread_config)
         self.computeengine = ComputeEngineConfig(thread_config=thread_config)
-        self.iam = IAMConfig(thread_config=thread_config)
+        self.iam = IAM(gcp_facade, iam_facade)
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/configs/services.py]: Same comment as with the Stack Driver Logging PR, I'd merge the two facades.

File: ScoutSuite/providers/gcp/facade/iam.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
+# -*- coding: utf-8 -*-
+
+from googleapiclient import discovery
+from ScoutSuite.providers.gcp.utils import MemoryCache
+
+class IAMFacade:
+    def __init__(self):
+        self._iam_client = discovery.build('iam', 'v1', cache_discovery=False, cache=MemoryCache())
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/facade/iam.py]: [Same comment](https://github.com/nccgroup/ScoutSuite/pull/229#discussion_r263377080), I don't think this is the right place to do HTTP requests.

File: ScoutSuite/providers/gcp/resources/iam/service_accounts.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    ]
+
+    def __init__(self, iam_facade, project_id):
+        self.iam_facade = iam_facade
+        self.project_id = project_id
+
+    async def fetch_all(self):
+        raw_service_accounts = await self.iam_facade.get_service_accounts(self.project_id)
+        for raw_service_account in raw_service_accounts:
+            sa_id, service_account = self._parse_service_account(raw_service_account)
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/resources/iam/service_accounts.py]: I'd rename `sa_id` for `service_account_id`.

File: ScoutSuite/providers/gcp/resources/iam/service_accounts.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+            self[sa_id] = service_account
+        await self._fetch_children()
+    
+    async def _fetch_children(self):
+        for sa_id, service_account in self.items():
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/resources/iam/service_accounts.py]: Same here, I'd rename `sa_id` to `service_account_id`.

File: ScoutSuite/providers/gcp/resources/iam/service_accounts.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            for child_name, child_class in self._children:
+                child = child_class(self.iam_facade, self.project_id, service_account['email'])
+                await child.fetch_all()
+                self[sa_id][child_name] = child
+
+    def _parse_service_account(self, raw_service_account):
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/resources/iam/service_accounts.py]: [Same as here](https://github.com/nccgroup/ScoutSuite/pull/229#discussion_r263382611), would rename to `_parse`.

File: ScoutSuite/providers/gcp/resources/iam/bindings.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+        self.project_id = project_id
+        self.service_account_email = service_account_email 
+
+    async def fetch_all(self):
+        raw_bindings = await self.iam_facade.get_bindings(self.project_id, self.service_account_email)
+        for raw_binding in raw_bindings.get('bindings', []):
+            binding_id, binding = self._parse_binding(raw_binding)
+            self[binding_id] = binding
+
+    def _parse_binding(self, raw_binding):
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/resources/iam/bindings.py]: [Same](https://github.com/nccgroup/ScoutSuite/pull/229#discussion_r263382611), would rename to `_parse`.

File: ScoutSuite/providers/gcp/resources/iam/keys.py [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+        self.project_id = project_id
+        self.service_account_email = service_account_email 
+
+    async def fetch_all(self):
+        raw_keys = await self.iam_facade.get_keys(self.project_id, self.service_account_email)
+        for raw_key in raw_keys.get('keys', []):
+            key_id, key = self._parse_key(raw_key)
+            self[key_id] = key
+
+    def _parse_key(self, raw_key):
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/resources/iam/keys.py]: [Same](https://github.com/nccgroup/ScoutSuite/pull/229#discussion_r263382611), would rename to `_parse`.

File: ScoutSuite/providers/gcp/facade/facade.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
+# -*- coding: utf-8 -*-
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/facade/facade.py]: [Same as here](https://github.com/nccgroup/ScoutSuite/pull/229#discussion_r263374292), would remove this line. I won't point out every occurrence but applies to all the other files.

File: ScoutSuite/providers/gcp/facade/facade.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+
+from googleapiclient import discovery
+from ScoutSuite.providers.gcp.utils import MemoryCache
+
+class GCPFacade:
+    def __init__(self):
+        self._resourcemanager_client = discovery.build('cloudresourcemanager', 'v1', cache_discovery=False, cache=MemoryCache())
+
+    # TODO: Make truly async    
+    async def get_projects(self):
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/facade/facade.py]: I think you could generalize this method. There is code duplication between this and [this](https://github.com/nccgroup/ScoutSuite/pull/230/files#diff-ffea74d1a65db135e82d200249fd7e5eR21).

**Comment by j4v** [comment]
> it seems like there's barely anyone using 3.5

:laughing: 

There's 2 types of people, those who install dependencies on their host system and burn in hell, and those who use virtual environments. The later shouldn't care what version of 3.x is used...

**Comment by Remi05** [comment]
> > > Do we still plan on supporting Python 3.5 or only 3.6 and 3.7?
> > 
> > 
> > Well asyncio is available from 3.5+, but 3.6+ should be the standard 3.x on most distros. If there's a benefit to only supporting 3.6+ that's something we could consider. I'm seeing a lot of changes happening in python 3.x so if we're along for the ride (and everyone will have to be from 2020 onward) then we might as well just accept that having outdated versions of python is not acceptable anymore.
> 
> I'm not sure formatted string literals is a good enough reason to deprecate 3.5... but at the same time, it seems like there's barely anyone using 3.5.
> 
> image

I wouldn't deprecated Python 3.5 over formatted strings, especially since you can format them differently, but if we were already going to, I would have kept this cleaner syntax 😉

File: ScoutSuite/providers/gcp/facade/facade.py (hunk 9 above)
Comment (Remi05) [reply ScoutSuite/providers/gcp/facade/facade.py]: Yeah, I had noticed but I wasn't sure how to do it cleanly, I think I found a decent way. Let me know what you think!

**Comment by Remi05** [comment]
I replaced formatted string using the `f'Text {variable}'` with `'{} text {}'.format(var1, var2)` to remain compatible with Python 3.5!

File: ScoutSuite/providers/gcp/facade/facade.py (hunk 9 above)
Comment (Aboisier) [reply ScoutSuite/providers/gcp/facade/facade.py]: Yup, that's exactly what I had in mind. You'll have to make this async though (like [this commit](https://github.com/nccgroup/ScoutSuite/pull/224/commits/25967e57b29d9c66af0daf339162aa5e0492e3ea)). Michael spent a few days looking for the performance bottleneck, and I spent a day, and it was that; it makes a huge difference.

File: ScoutSuite/providers/gcp/facade/facade.py [hunk 10]
```diff
@@ -0,0 +1,12 @@
+
+class Facade:
```
Comment (misg) [inline ScoutSuite/providers/gcp/facade/facade.py]: Is this an attempt to make an abstract interface for all cloud provider facades? Seeing how really different they are, I'm not sure we can really make them homogeneous...

File: ScoutSuite/providers/gcp/facade/iam.py [hunk 11]
```diff
@@ -0,0 +1,30 @@
+from googleapiclient import discovery
+from ScoutSuite.providers.gcp.utils import MemoryCache
+from ScoutSuite.providers.gcp.facade.facade import Facade
+from ScoutSuite.providers.gcp.facade.utils import GCPFacadeUtils
+
```
Comment (misg) [inline ScoutSuite/providers/gcp/facade/iam.py]:
```suggestion
from ScoutSuite.providers.utils import run_concurrently

```

File: ScoutSuite/providers/gcp/facade/iam.py [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+class IAMFacade(Facade):
+    def _build_client(self):
+        return discovery.build('iam', 'v1', cache_discovery=False, cache=MemoryCache())
+
+    # TODO: Make truly async
+    async def get_bindings(self, project_id, service_account_email):
+        resource = 'projects/{}/serviceAccounts/{}'.format(project_id, service_account_email)
+        iam_client = self._get_client()
+        response = iam_client.projects().serviceAccounts().getIamPolicy(resource=resource).execute()
```
Comment (misg) [inline ScoutSuite/providers/gcp/facade/iam.py]:
```suggestion
        response = await run_concurrently(
                lambda: iam_client.projects().serviceAccounts().getIamPolicy(resource=resource).execute()
        )
```

File: ScoutSuite/providers/gcp/facade/iam.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+        return response.get('bindings', [])
+
+    # TODO: Make truly async
+    async def get_keys(self, project_id, service_account_email):
+        name = 'projects/{}/serviceAccounts/{}'.format(project_id, service_account_email)
+        iam_client = self._get_client()
+        response = iam_client.projects().serviceAccounts().keys().list(name=name).execute()
```
Comment (misg) [inline ScoutSuite/providers/gcp/facade/iam.py]:
```suggestion
        response = await run_concurrently(
                lambda: iam_client.projects().serviceAccounts().keys().list(name=name).execute()
        )
```

File: ScoutSuite/providers/gcp/facade/iam.py [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+        return response.get('keys', [])
+
+    # TODO: Make truly async
```
Comment (misg) [inline ScoutSuite/providers/gcp/facade/iam.py]: Already truly async afaik, unless there's an API call behind the assignment of `request` and/or `service_accounts_group`?

File: ScoutSuite/providers/gcp/resources/iam/service_accounts.py [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+
+    async def fetch_all(self):
+        raw_service_accounts = await self.gcp_facade.iam.get_service_accounts(self.project_id)
+        for raw_service_account in raw_service_accounts:
+            service_account_id, service_account = self._parse_service_account(raw_service_account)
+            self[service_account_id] = service_account
+        await self._fetch_children()
+    
+    async def _fetch_children(self):
+        for service_account_id, service_account in self.items():
```
Comment (misg) [inline ScoutSuite/providers/gcp/resources/iam/service_accounts.py]: You'll need to parallelize this loop (see https://github.com/nccgroup/ScoutSuite/pull/229/files#r263374926) and probably make a refactoring of `fetch_children` as well (would be probably worth it to define a `GCPCompositeResources` to do so, like `AWSCompositeResources` and `AzureCompositeResources` ([here](https://github.com/nccgroup/ScoutSuite/blob/28e6d1f3e4bd4b068c0b0caf31eb718743919d3c/ScoutSuite/providers/aws/resources/resources.py#L25) and [there](https://github.com/nccgroup/ScoutSuite/blob/28e6d1f3e4bd4b068c0b0caf31eb718743919d3c/ScoutSuite/providers/azure/resources/resources.py#L8))).

File: ScoutSuite/providers/gcp/facade/iam.py (hunk 14 above)
Comment (Remi05) [reply ScoutSuite/providers/gcp/facade/iam.py]: That was before I extracted the `get_all()` loop, I'll update the comment 😉

File: ScoutSuite/providers/gcp/facade/facade.py (hunk 9 above)
Comment (Remi05) [reply ScoutSuite/providers/gcp/facade/facade.py]: I'm aware, that's what the comment was for, I was waiting for Michael's PR to be merged before adding it! 😉

File: ScoutSuite/providers/gcp/facade/facade.py (hunk 10 above)
Comment (Remi05) [reply ScoutSuite/providers/gcp/facade/facade.py]: No it was simply for GCP, it avoids duplicating the `_get_client()` method, I can give it another name though

File: ScoutSuite/providers/gcp/resources/iam/service_accounts.py (hunk 15 above)
Comment (Remi05) [reply ScoutSuite/providers/gcp/resources/iam/service_accounts.py]: Done!

File: ScoutSuite/providers/gcp/facade/facade.py (hunk 10 above)
Comment (Aboisier) [reply ScoutSuite/providers/gcp/facade/facade.py]: For AWS, I called my base facade  `AWSBaseFacade`. I think it's less confusing than `Facade`.

File: ScoutSuite/providers/gcp/facade/iam.py [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+        iam_client = self._get_client()
+        response = await run_concurrently(
+                lambda: iam_client.projects().serviceAccounts().keys().list(name=name).execute()
+        )
+        return response.get('keys', [])
+
+    async def get_service_accounts(self, project_id):
+        name = 'projects/{}'.format(project_id)     
+        iam_client = self._get_client()
+        request = iam_client.projects().serviceAccounts().list(name=name)
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/facade/iam.py]: I think this should be parallelized!

```suggestion
        request = await run_concurrently(
                iam_client.projects().serviceAccounts().list(name=name)
        )
```

File: ScoutSuite/providers/gcp/facade/iam.py [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+        service_accounts_group = iam_client.projects().serviceAccounts()
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/facade/iam.py]: Should that be parallelized too?
```suggestion
        service_accounts_group = await run_concurrently(
                iam_client.projects().serviceAccounts()
        )
```

File: ScoutSuite/providers/gcp/facade/iam.py (hunk 16 above)
Comment (Remi05) [reply ScoutSuite/providers/gcp/facade/iam.py]: `request` and `services_accounts_group` simply build the request from my understanding. The actual fetching is done in `GCPFacadeUtils.get_all()` which is already parallelized .

File: ScoutSuite/providers/gcp/facade/facade.py (hunk 10 above)
Comment (Remi05) [reply ScoutSuite/providers/gcp/facade/facade.py]: Renamed it to `GCPBaseFacade`!

**Comment by Remi05** [comment]
Fixed UI and added left menu for projects.

Before : 

iam_before

After : 

iam_after

File: ScoutSuite/providers/gcp/resources/iam/keys.py [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+
+    async def fetch_all(self):
+        raw_keys = await self.gcp_facade.iam.get_keys(self.project_id, self.service_account_email)
+        for raw_key in raw_keys:
+            key_id, key = self._parse_key(raw_key)
+            self[key_id] = key
+
+    def _parse_key(self, raw_key):
+        key_dict = {}
+        key_dict['id'] = raw_key['name'].split('/')[-1]
```
Comment (Aboisier) [inline ScoutSuite/providers/gcp/resources/iam/keys.py]: Can we add a small comment here explaining where this comes from?

Comment (Remi05) [reply ScoutSuite/providers/gcp/resources/iam/keys.py]: Done!

**Pull request merged**